### PR TITLE
Adding mutual TLS with ID Secret authenticator

### PR DESCRIPTION
**Pull request by ChamathNS**
## Purpose
$subject
Introduce a property to mandate mutual SSL
porting from https://github.com/wso2-extensions/identity-oauth-addons/pull/25
porting from https://github.com/wso2-extensions/identity-oauth-addons/pull/28

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-scr-plugin</artifactId>
+                <version>${maven.source.plugin.version}</version>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>${maven.surefire.plugin.version}</version>
+                <configuration>
+                    <!--<argLine>-Xmx1024m -XX:PermSize=256m -XX:MaxPermSize=512m -Xdebug -Xnoagent -Djava.compiler=NONE -Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=5006</argLine>-->
```
Comment (pulasthi7) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml:105]: seems unwanted line

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                        </goals>
+                        <configuration>
+                            <rules>
+                                <rule implementation="org.jacoco.maven.RuleConfiguration">
+                                    <element>BUNDLE</element>
+                                    <limits>
+                                        <limit implementation="org.jacoco.report.check.Limit">
+                                            <counter>COMPLEXITY</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <!--<minimum>0.4</minimum>-->
```
Comment (pulasthi7) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml:157]: can we have this uncommented?

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/internal/MutualTLSServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            bundleContext = context.getBundleContext();
+            MutualTLSWithIdSecretAuthenticator mutualTLSWithIdSecretAuthenticator =
+                    new MutualTLSWithIdSecretAuthenticator();
+            bundleContext.registerService(OAuthClientAuthenticator.class.getName(), mutualTLSWithIdSecretAuthenticator,
+                    null);
+            if (log.isDebugEnabled()) {
+                log.debug("Mutual TLS with basic auth bundle is activated");
+            }
+
+        } catch (Throwable e) {
```
Comment (pulasthi7) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/internal/MutualTLSServiceComponent.java:56]: catch exception instead

Comment (ChamathNS) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/internal/MutualTLSServiceComponent.java]: Addressed with https://github.com/wso2-extensions/identity-oauth-addons/pull/69/commits/a326ec1044f2fd02b360fef4de21ce4a346ec808.

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml (hunk 2 above)
Comment (ChamathNS) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml]: Addressed with https://github.com/wso2-extensions/identity-oauth-addons/pull/69/commits/a326ec1044f2fd02b360fef4de21ce4a346ec808

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml (hunk 1 above)
Comment (ChamathNS) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml]: Addressed with https://github.com/wso2-extensions/identity-oauth-addons/pull/69/commits/a326ec1044f2fd02b360fef4de21ce4a346ec808

**Review by pulasthi7 — APPROVED** [review]

**Pull request merged**